common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,198 @@
+    }
+  }
+
+  /**
+   * Finish a Trace associated with given transactionId.
+   *
+   * @param transactionId
+   */
+  def finish(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+    if (enabled)
+      clear(transactionId)
+  }
+
+  /**
+   * Register error
+   *
+   * @param transactionId
+   */
+  def error(transactionId: TransactionId): Unit = {
+    if (enabled)
+      clear(transactionId)
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = {
+    var contextMap: Option[Map[String, String]] = None
+    if (enabled) {
+      spanMap.get(transactionId.meta.id) match {
+        case Some(spanList) => {
+          var map: java.util.Map[String, String] = new java.util.HashMap()
+          //inject latest span context in map
+          GlobalTracer.get().inject(spanList.last.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(map))
+          contextMap = Some(scala.collection.JavaConverters.mapAsScalaMapConverter(map).asScala.toMap)
+        }
+        case None => None
+      }
+    }
+    contextMap
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]) = {
+    if (enabled) {
+      context match {
+        case Some(scalaMap) => {
+          var javaMap: java.util.Map[String, String] =
+            scala.collection.JavaConverters.mapAsJavaMapConverter(scalaMap).asJava
+          var ctx: SpanContext = GlobalTracer.get().extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(javaMap))
+          contextMap.put(transactionId.meta.id, ctx)
+        }
+        case None =>
+      }
+    }
+  }
+
+  def clear(transactionId: TransactionId): Unit = {
+    spanMap.get(transactionId.meta.id) match {
+      case Some(spanList) => {
+        spanList.last.deactivate()
+        spanList.remove(spanList.size - 1)
+        if (spanList.isEmpty) {
+          spanMap.remove(transactionId.meta.id)
+          contextMap.remove(transactionId.meta.id)
+        }
+
+      }
+      case None =>
+    }
+
+  }
+
+  def configureTracer(componentName: String): Unit = {
+    val tracingConfig = loadConfigOrThrow[TracingConfig](ConfigKeys.tracing)
+
+    enabled = tracingConfig.enabled
+    if (enabled) {
+      val tracer = tracingConfig.tracer
+
+      if (tracer.equalsIgnoreCase("zipkin")) {
+        //configure opentracing with zipkin
+        val sender: Sender = OkHttpSender.create(
+          "http://" +
+            tracingConfig.zipkin.get("host").getOrElse("localhost") + ":" +
+            tracingConfig.zipkin.get("port").getOrElse(9411) + "/api/v1/spans")
+
+        val reporter: Reporter[zipkin.Span] = AsyncReporter.builder(sender).build()
```

Comment:
Instead of using generic key-value map in config better to use actual key names. So above flow can be simplified to. May be another config for protocol also is needed

```scala
case class TracingConfig(enabled: Boolean, component: String, zipkin: ZipkinConfig)
case class ZipkinConfig(host: String, port: Int) {
  def url = s"http://$host:$port//api/v1/spans"
}
```